contracts/TokenMigration.aes

```diff
@@ -14,46 +19,50 @@ payable contract TokenMigration =
-        require(is_contained, "From provided data, cannot be generated same root")
+        let ae_address_str = Address.to_str(ae_address)
+        let unhashed_msg = String.concat("\x19Ethereum Signed Message:\n", String.concat(Int.to_str(String.length(ae_address_str)), ae_address_str))
+        let hashed_msg = String.sha3(unhashed_msg)
 
-        // extract signer of signature, should be same address that holds tokens
-        let recovered_address = get_signer(msg_hash, sig)
-        let are_equal = switch (recovered_address)
-                            None => false
-                            Some(recovered) => recovered == eth_address
+        // By providing the hashed_ae_address to the ecrevocer we are sure that the eth_address is correct
+        let recovered_eth_address = get_signer(hashed_msg, sig)
+        let recovered_eth_address_string = switch (recovered_eth_address)
+                                            None => abort("Failed to recover address, bad signature")
+                                            Some(recovered) => String.concat("0X", Bytes.to_str(recovered))
 
-        require(are_equal, "Mismatch between passed eth address and recovered one")
+        require(!is_migrated(recovered_eth_address_string), "This account has already transferred its tokens")
+        require(contained_in_merkle_tree(recovered_eth_address_string, token_amount, leaf_index, siblings), "From provided data, cannot be generated same root")
 
-        Chain.spend(ae_address, amount_of_tokens)
+        Chain.spend(ae_address, token_amount)
         put(state{migrations_count = state.migrations_count + 1})
-        put(state{map_transfered_accounts[eth_addr_str] = ()})
+        put(state{map_transfered_accounts[recovered_eth_address_string] = ()})
 
-        Chain.event(Migrated(eth_address, ae_address, amount_of_tokens))
+        Chain.event(Migrated(recovered_eth_address_string, ae_address, token_amount))
 
         state.migrations_count
 
+
     function get_signer(msg_hash: hash, sig: bytes(65)) =
         Crypto.ecrecover_secp256k1(msg_hash, sig)
 
-    entrypoint contained_in_merkle_tree(eth_addr: string, tokens: int, leaf_index: int, siblings: list(string) ) =
-        // concat data => eth_Addr:tokens => concatenated data
-        let data = String.concat(eth_addr, ":")
-        let data = String.concat(data, Int.to_str(tokens))
+    entrypoint contained_in_merkle_tree(eth_addr: string, token_amount: int, leaf_index: int, siblings: list(string) ) =
+        // concat leaf_data => eth_Addr:tokens => concatenated data
+        let leaf_data = String.concat(eth_addr, String.concat(":", Int.to_str(token_amount)))
 
         // concatenated data -> hash
-        let data = Bytes.to_str(String.sha3(data)) // hash -> str
-
-        //let data = Bytes.to_str(String.sha3(String.concat(eth_addr, ":", tokens))) // not working ?!
+        let leaf_data_hashed = Bytes.to_str(String.sha3(leaf_data)) // hash -> str
```

Comment:
Consider working with bytes instead of strings. `Crypto.sha3` hashes the raw byte array for `bytes` and we now have `Bytes.concat` which will let you do the calculation with root and siblings all being `bytes(32)`.

Reply:
I'd agree, staying with bytes would be "nicer"